File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.cpp
// TC- O(n)
// SC- O(1)

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode* dummy = new ListNode(0, head);
        ListNode* slowptr = dummy;
        ListNode* fastptr = dummy;
        for(int i=0; i<n+1; i++)
            fastptr = fastptr->next;
        while(fastptr){
            fastptr = fastptr->next;
            slowptr = slowptr->next;
        }
        slowptr->next = slowptr->next->next;
        return dummy->next;
    }
};

File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.py
# TC- O(n)
# SC- O(1)

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummy = ListNode()
        dummy.next = head
        slowptr = dummy
        fastptr = dummy

        while n>=0:
            fastptr = fastptr.next
            n -= 1
        
        while fastptr:
            slowptr = slowptr.next
            fastptr = fastptr.next

        slowptr.next = slowptr.next.next

        return dummy.next